File: android/app/src/main/java/com/mgbridge/companion/net/Messages.kt
package com.mgbridge.companion.net

import org.json.JSONArray
import org.json.JSONObject

/**
 * Control frames of the bridge protocol (docs/PROTOCOL.md). Every frame is a JSON
 * object with a `t` discriminator; this file is pure JVM so the codec is unit-testable
 * off-device, and the Swift side mirrors it against the same test vectors.
 */

const val PROTOCOL_VERSION = 1

data class FileMeta(val name: String, val size: Long, val mime: String)

sealed class Frame {
    data class Hello(val v: Int, val name: String) : Frame()
    data class PairReq(val name: String, val proof: String) : Frame()
    data class PairOk(val name: String) : Frame()
    data class PairErr(val reason: String) : Frame()
    data class Offer(val files: List<FileMeta>) : Frame()
    data object Accept : Frame()
    data class Reject(val reason: String) : Frame()
    /** Announces that exactly `files[i].size` raw bytes follow this frame. */
    data class FileHeader(val i: Int) : Frame()
    data class Done(val sha256: List<String>) : Frame()
    data class Receipt(val ok: List<Boolean>) : Frame()
    data class Clip(val text: String) : Frame()
    data object Bye : Frame()
    data class Err(val reason: String) : Frame()

    fun toJson(): JSONObject {
        val o = JSONObject()
        when (this) {
            is Hello -> o.put("t", "hello").put("v", v).put("name", name)
            is PairReq -> o.put("t", "pairReq").put("name", name).put("proof", proof)
            is PairOk -> o.put("t", "pairOk").put("name", name)
            is PairErr -> o.put("t", "pairErr").put("reason", reason)
            is Offer -> {
                val arr = JSONArray()
                for (f in files) arr.put(
                    JSONObject().put("name", f.name).put("size", f.size).put("mime", f.mime)
                )
                o.put("t", "offer").put("files", arr)
            }
            is Accept -> o.put("t", "accept")
            is Reject -> o.put("t", "reject").put("reason", reason)
            is FileHeader -> o.put("t", "file").put("i", i)
            is Done -> o.put("t", "done").put("sha256", JSONArray(sha256))
            is Receipt -> o.put("t", "receipt").put("ok", JSONArray(ok))
            is Clip -> o.put("t", "clip").put("text", text)
            is Bye -> o.put("t", "bye")
            is Err -> o.put("t", "err").put("reason", reason)
        }
        return o
    }

    companion object {
        /** @throws org.json.JSONException on missing fields or an unknown `t`. */
        fun fromJson(o: JSONObject): Frame = when (val t = o.getString("t")) {
            "hello" -> Hello(o.getInt("v"), o.getString("name"))
            "pairReq" -> PairReq(o.getString("name"), o.getString("proof"))
            "pairOk" -> PairOk(o.getString("name"))
            "pairErr" -> PairErr(o.getString("reason"))
            "offer" -> {
                val arr = o.getJSONArray("files")
                Offer((0 until arr.length()).map { idx ->
                    val f = arr.getJSONObject(idx)
                    FileMeta(f.getString("name"), f.getLong("size"), f.getString("mime"))
                })
            }
            "accept" -> Accept
            "reject" -> Reject(o.getString("reason"))
            "file" -> FileHeader(o.getInt("i"))
            "done" -> {
                val arr = o.getJSONArray("sha256")
                Done((0 until arr.length()).map { arr.getString(it) })
            }
            "receipt" -> {
                val arr = o.getJSONArray("ok")
                Receipt((0 until arr.length()).map { arr.getBoolean(it) })
            }
            "clip" -> Clip(o.getString("text"))
            "bye" -> Bye
            "err" -> Err(o.getString("reason"))
            else -> throw org.json.JSONException("unknown frame type: $t")
        }
    }
}

File: macos/Sources/mgbridge/Messages.swift
import Foundation

/// Control frames of the bridge protocol (docs/PROTOCOL.md). Mirrors the Kotlin codec
/// and is tested against the same vectors. JSONSerialization (not Codable) because a
/// `t`-discriminated union maps to it more directly.

struct FileMeta: Equatable {
    let name: String
    let size: Int64
    let mime: String
}

enum FrameError: Error, CustomStringConvertible {
    case malformed(String)

    var description: String {
        switch self {
        case .malformed(let why): return "malformed frame: \(why)"
        }
    }
}

enum Frame: Equatable {
    static let protocolVersion = 1

    case hello(v: Int, name: String)
    case pairReq(name: String, proof: String)
    case pairOk(name: String)
    case pairErr(reason: String)
    case offer(files: [FileMeta])
    case accept
    case reject(reason: String)
    /// Announces that exactly `files[i].size` raw bytes follow this frame.
    case file(i: Int)
    case done(sha256: [String])
    case receipt(ok: [Bool])
    case clip(text: String)
    case bye
    case err(reason: String)

    func jsonObject() -> [String: Any] {
        switch self {
        case .hello(let v, let name):
            return ["t": "hello", "v": v, "name": name]
        case .pairReq(let name, let proof):
            return ["t": "pairReq", "name": name, "proof": proof]
        case .pairOk(let name):
            return ["t": "pairOk", "name": name]
        case .pairErr(let reason):
            return ["t": "pairErr", "reason": reason]
        case .offer(let files):
            return ["t": "offer", "files": files.map { f -> [String: Any] in
                ["name": f.name, "size": f.size, "mime": f.mime]
            }]
        case .accept:
            return ["t": "accept"]
        case .reject(let reason):
            return ["t": "reject", "reason": reason]
        case .file(let i):
            return ["t": "file", "i": i]
        case .done(let hashes):
            return ["t": "done", "sha256": hashes]
        case .receipt(let ok):
            return ["t": "receipt", "ok": ok]
        case .clip(let text):
            return ["t": "clip", "text": text]
        case .bye:
            return ["t": "bye"]
        case .err(let reason):
            return ["t": "err", "reason": reason]
        }
    }

    func encoded() throws -> Data {
        try JSONSerialization.data(withJSONObject: jsonObject(), options: [.sortedKeys])
    }

    static func decode(_ data: Data) throws -> Frame {
        guard let obj = try? JSONSerialization.jsonObject(with: data),
              let dict = obj as? [String: Any] else {
            throw FrameError.malformed("not a JSON object")
        }
        return try from(json: dict)
    }

    static func from(json obj: [String: Any]) throws -> Frame {
        func str(_ key: String) throws -> String {
            guard let s = obj[key] as? String else { throw FrameError.malformed("missing \(key)") }
            return s
        }
        func int(_ key: String) throws -> Int {
            guard let n = obj[key] as? NSNumber else { throw FrameError.malformed("missing \(key)") }
            return n.intValue
        }
        guard let t = obj["t"] as? String else { throw FrameError.malformed("missing t") }
        switch t {
        case "hello":
            return .hello(v: try int("v"), name: try str("name"))
        case "pairReq":
            return .pairReq(name: try str("name"), proof: try str("proof"))
        case "pairOk":
            return .pairOk(name: try str("name"))
        case "pairErr":
            return .pairErr(reason: try str("reason"))
        case "offer":
            guard let arr = obj["files"] as? [[String: Any]] else {
                throw FrameError.malformed("missing files")
            }
            let files = try arr.map { f -> FileMeta in
                guard let name = f["name"] as? String,
                      let size = f["size"] as? NSNumber,
                      let mime = f["mime"] as? String else {
                    throw FrameError.malformed("bad file entry")
                }
                return FileMeta(name: name, size: size.int64Value, mime: mime)
            }
            return .offer(files: files)
        case "accept":
            return .accept
        case "reject":
            return .reject(reason: try str("reason"))
        case "file":
            return .file(i: try int("i"))
        case "done":
            guard let hashes = obj["sha256"] as? [String] else {
                throw FrameError.malformed("missing sha256")
            }
            return .done(sha256: hashes)
        case "receipt":
            guard let ok = obj["ok"] as? [Bool] else {
                throw FrameError.malformed("missing ok")
            }
            return .receipt(ok: ok)
        case "clip":
            return .clip(text: try str("text"))
        case "bye":
            return .bye
        case "err":
            return .err(reason: try str("reason"))
        default:
            throw FrameError.malformed("unknown frame type \(t)")
        }
    }
}
